compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -48,6 +51,46 @@ use rustc_hir::intravisit::Visitor;
 use std::cmp::{self, Ordering};
 use std::iter;
 
+#[derive(Debug)]
+struct ConfusablesDirective {
+    names: Vec<String>,
+}
+
+#[derive(Debug)]
+struct ConfusablesDirectiveParseErr {
```

Comment:
this should be a session diagnostics like the others in `rustc_hir_typeck/src/errors.rs` (see https://rustc-dev-guide.rust-lang.org/diagnostics/translation.html#writing-translatable-diagnostics)

this way you can easily return it and then just call `sess.emit_err` in the caller.

Reply:
You mean something like

```rust
#[derive(Diagnostic)]
#[diag(hir_typeck_invalid_rustc_confusable_attr)]
struct ConfusablesDirectiveParseErr {
    #[primary_span]
    #[label]
    span: Span,
}
```

right?

Reply:
yes